imports/wasi_snapshot_preview1/poll.go

```diff
@@ -72,88 +80,160 @@ func pollOneoffFn(ctx context.Context, mod api.Module, params []uint64) syscall.
 	for i := uint32(0); i < nsubscriptions; i++ {
 		inOffset := i * 48
 		outOffset := i * 32
 
 		eventType := inBuf[inOffset+8] // +8 past userdata
-		var errno syscall.Errno        // errno for this specific event (1-byte)
+		// +8 past userdata +8 contents_offset
+		argBuf := inBuf[inOffset+8+8:]
+		userData := inBuf[inOffset : inOffset+8]
+
+		evt := &event{
+			eventType: eventType,
+			userData:  userData,
+			errno:     wasip1.ErrnoSuccess,
+			outOffset: outOffset,
+		}
+
 		switch eventType {
 		case wasip1.EventTypeClock: // handle later
-			// +8 past userdata +8 contents_offset
-			errno = processClockEvent(ctx, mod, inBuf[inOffset+8+8:])
-		case wasip1.EventTypeFdRead, wasip1.EventTypeFdWrite:
-			// +8 past userdata +8 contents_offset
-			errno = processFDEvent(mod, eventType, inBuf[inOffset+8+8:])
+			newTimeout, err := processClockEvent(argBuf)
+			if err != 0 {
+				return err
+			}
+			// Min timeout.
+			if newTimeout < timeout {
+				timeout = newTimeout
+			}
+			// Ack the clock event to the outBuf.
+			writeEvent(outBuf, evt)
+		case wasip1.EventTypeFdRead:
+			fd := le.Uint32(argBuf)
+			if fd == internalsys.FdStdin {
+				// if the fd is Stdin, do not ack yet,
+				// append to a slice for delayed evaluation.
+				stdinSubs = append(stdinSubs, evt)
+			} else {
+				evt.errno = processFDEventRead(fsc, fd)
+				writeEvent(outBuf, evt)
+				readySubs++
+			}
+		case wasip1.EventTypeFdWrite:
+			fd := le.Uint32(argBuf)
+			evt.errno = processFDEventWrite(fsc, fd)
+			readySubs++
+			writeEvent(outBuf, evt)
 		default:
 			return syscall.EINVAL
 		}
+	}
+
+	// If there are subscribers with data ready, we have already written them to outBuf,
+	// and we don't need to wait for the timeout: clear it.
+	if readySubs != 0 {
+		timeout = 0
+	}
 
-		// Write the event corresponding to the processed subscription.
-		// https://github.com/WebAssembly/WASI/blob/snapshot-01/phases/snapshot/docs.md#-event-struct
-		copy(outBuf, inBuf[inOffset:inOffset+8]) // userdata
-		if errno != 0 {
-			outBuf[outOffset+8] = byte(wasip1.ToErrno(errno)) // uint16, but safe as < 255
-		} else { // special case ass ErrnoSuccess is zero
-			outBuf[outOffset+8] = 0
+	// If there are stdin subscribers, check for data with given timeout.
+	if len(stdinSubs) > 0 {
+		reader := getStdioFileReader(mod)
+		// Wait for the timeout to expire, or for some data to become available on Stdin.
+		stdinReady, err := reader.Poll(timeout)
+		if err != nil {
+			return platform.UnwrapOSError(err)
 		}
-		outBuf[outOffset+9] = 0
-		le.PutUint32(outBuf[outOffset+10:], uint32(eventType))
-		// TODO: When FD events are supported, write outOffset+16
+		if stdinReady {
+			// stdin has data ready to for reading, write back all the events
+			for i := range stdinSubs {
+				evt := stdinSubs[i]
+				evt.errno = 0
+				writeEvent(outBuf, evt)
+			}
+		}
+	} else {
+		// No subscribers, just wait for the given timeout.
+		time.Sleep(timeout)
```

Comment:
I think this should be back to `sysCtx.Nanosleep()` but if I just revert this line naively a bunch of tests will break, so I will investigate this further

Reply:
We should also probably select on context cancellation in case the program wants to interrupt the blocking call?
```go
timer := time.NewTimer(timeout)
defer timer.Stop()
select {
case <-timer.C:
case <-ctx.Done():
}
```
It might be tricky to make context cancellation work with `platform.Select` as well.

Reply:
yeah it will be tricky to avoid leaking the underlying clock if/when a fake impl can work.

Reply:
If we want to support context cancellation, a common approach is to add a _signal_ file descriptor to the set, e.g. the read-end of a pipe or an eventfd on linux. When the module is closed, maybe asynchronously by the goroutine watching for context cancellation, we unblock the `platform.Select` call by writing to the fd.

Reply:
works for me! cool thing is because WASI doesn't support signals we don't need to worry about conflict interpreting them ;)

That said, this will need to be a part of the platform api sig not wasi, as it is an implementation detail of how we are supporting the context param.

So we'll have to special case the function signature or its impl, to support that. You might want to give some more concrete hints in the go side (code snippet) if you have some in mind.

Thanks for the idea!

Reply:
I think we're going to quickly hit the case where someone runs into the limitations of the current implementation, but it's OK to keep it for a follow up 👍